src/Http/CurlClient.php

```diff
@@ -0,0 +1,207 @@
+     */
+    function doGET(String $url, $headers = []) : ResponseInterface
+    {
+        // Set GET as the method with an empty body.
+        return $this->doCurl($url, 'GET', '', $headers);
+    }
+
+
+    /**
+     * Generic curl POST request.  Builds the cURL handler and sets the options
+     * corresponding to a POST request.
+     *
+     * @param string $url       The resource to POST to.
+     * @param mixed  $post_body An array or string for the POST request body.
+     * @param array  $headers   Misc additional HTTP headers
+     *
+     * @return ResponseInterface The HTTP response to the POST request.
+     */
+    function doPOST(String $url, $post_body, $headers = []) : ResponseInterface
+    {
+        return $this->doCurl((string) $url, 'POST', $post_body, $headers);
+    }
+
+    /**
+     * Generic curl object. Currently handles cases for GET and POST with HEAD
+     * partially implemented in case the LORIS API ends up using it soon.
+     * Opens, configures, executes, and closes a cURL object.
+     *
+     * @param string $url       The resource to request.
+     * @param string $method    The HTTP method. GET and POST supported.
+     * @param mixed  $post_body An array or string for the request body.
+     * @param array  $headers   Misc additional HTTP headers
+     *
+     * @return ResponseInterface The HTTP response to the request.
+     */
+    function doCurl(
+        String $url,
+        String $method,
+        $post_body,
+        Array $headers = []
+    ) : ResponseInterface {
+
+        /* Build curl and set options */
+        $ch = curl_init();
+        curl_setopt($ch, CURLOPT_URL, $url);
+        curl_setopt($ch, CURLOPT_HEADER, true);
+
+        // Alter curl structure based on HTTP method.
+        if ($method === 'POST') {
+            /* POST body */
+            if (empty($post_body)) {
+                throw new \Exception(
+                    "Method selected is POST but body is empty!"
+                );
+            }
+            curl_setopt($ch, CURLOPT_POSTFIELDS, $post_body);
+            curl_setopt($ch, CURLOPT_POST, 1);
+        } else if ($method === 'HEAD') { // TODO: This isn't actually implemented yet
+            curl_setopt($ch, CURLOPT_NOBODY, true); // read: 'no body'
+        }
+        // Follow redirects.
+        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
+        curl_setopt($ch, CURLOPT_VERBOSE, false);
+        // Capture response instead of printing it.
+        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+        // Connect via HTTP/2 if available. Some versions of PHP/cURL don't
+        // know about it so we are using the magic number 3 here.
+        // The usual alias is CURL_HTTP_VERSION_2_0 but Travis has an issue with
+        // it.
+        curl_setopt($ch, CURLOPT_HTTP_VERSION, 3);
+
+        // For VMs with expired ssl certificate change the value to false.
+        curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, true);
```

Comment:
We don't currently depend on the curl PECL extension, so this will break on a lot of LORIS installs.. especially if it's not declared as required in the composer.json file.

Reply:
I think we can use apt instead of pecl, not sure if that matters. 

I can also include this line in the automatic update script -- if we plan to use it in the future then it could help ameliorate this issue.

I'm also hoping to use cURL for this PR https://github.com/aces/Loris/pull/3961 to interact with the Have I Been Pwned API.

Are you opposed to including curl as a requirement in future versions?

Reply:
It doesn't matter how it's installed, I just mean that it needs to declared in composer.json if it's depended on.

I've started working on #3970 (both because it's long overdue and because I need to be able to use the API for real under my dev environment..). With that, I don't think you need curl at all. You can get the ResponseInterface by simply calling the handler function directly on the router. You'd get a similar level of confidence but eliminate the network overhead, process forking overhead, and dependency on a network to run your test.

Reply:
We need something that can send an HTTP request. I need it for cbrain api calls. @driusan Would adding `ext-curl` to the required package be sufficient?